File: python/trie.py
class TrieNode: 
	
    # Trie node class 
    def __init__(self): 
            self.children = [None]*5

            # isEndOfWord is True if node represent the end of the word 
            self.isEndOfWord = False

class Trie: 
	
    # Trie data structure class 
    def __init__(self): 
            self.root = self.getNode() 

    def getNode(self): 
    
            # Returns new trie node (initialized to NULLs) 
            return TrieNode() 

    def _charToIndex(self,ch): 
            
            # private helper function 
            # Converts key current character into index 
            # use only 'a' through 'z' and lower case 
            chr_to_index = {'A' : 0, 'C' : 1, 'G' : 2, 'T' : 3, 'N' : 4}            
            return chr_to_index[ch]


    def insert(self,key): 
            
            # If not present, inserts key into trie 
            # If the key is prefix of trie node, 
            # just marks leaf node 
            pCrawl = self.root 
            length = len(key) 
            for level in range(length): 
                    index = self._charToIndex(key[level]) 

                    # if current character is not present 
                    if not pCrawl.children[index]: 
                            pCrawl.children[index] = self.getNode() 
                    pCrawl = pCrawl.children[index] 

            # mark last node as leaf 
            pCrawl.isEndOfWord = True

    def search(self, key): 
            
            # Search key in the trie 
            # Returns true if key presents 
            # in trie, else false 
            pCrawl = self.root 
            length = len(key) 
            for level in range(length): 
                    index = self._charToIndex(key[level]) 
                    if not pCrawl.children[index]: 
                            return False
                    pCrawl = pCrawl.children[index] 

            return pCrawl != None and pCrawl.isEndOfWord 


def generate_kmers(infile, kmer_size):
    with open(infile) as in_fh:
        while in_fh:
            in_fh.readline()
            read = in_fh.readline().strip()
            in_fh.readline()
            in_fh.readline()

            for i in range(len(read) - kmer_size + 1):
                kmer = read[i : i + kmer_size]
                yield kmer


# driver function 
def main(): 

    infile = "/Users/dongkim/Downloads/SRR062634.filt.fastq"
    kmer_size = 50

    # Trie object 
    t = Trie() 

    # Construct trie 
    for kmer in generate_kmers(infile, kmer_size): 
        t.insert(kmer) 


if __name__ == '__main__': 
	main() 

# This code is contributed by Atul Kumar (www.facebook.com/atul.kr.007) 

File: java/Trie.java
public class Trie {
    
    static final int ALPHABET_SIZE = 26;

    static class TrieNode { 
    
        TrieNode[] children = new TrieNode[ALPHABET_SIZE];
        boolean isEndOfWord;

        TrieNode() {
            isEndOfWord = false;
            for (int i = 0; i < ALPHABET_SIZE; i++)
                children[i] = null;
        }
    };

    static TrieNode root;

    static void insert(String key) {
        int level;
        int length = key.length();
        int index;

        TrieNode cur = root;

        for (int i = 0; i < length; i++) {
            index = key.charAt(i) - 'a';
            if (cur.children[index] == null)
                cur.children[index] = new TrieNode();
            cur = cur.children[index];
        }
        cur.isEndOfWord = true;
    }

    static boolean search(String key) {
        int length = key.length();
        int index;

        TrieNode cur = root;

        for (int i = 0; i  < length; i++) {
            index = key.charAt(i) - 'a';
            if (cur.children[index] == null)
                return false;
            cur = cur.children[index];
        }

        return (cur != null && cur.isEndOfWord);
    }

    static boolean isEmpty() {
        for (int i = 0; i < ALPHABET_SIZE; i++) {
            if (root.children[i] != null)
                return false;
        }

        return true;
    }

    public static void main(String args[])
    {
        // Input keys (use only 'a' through 'z' and lower case)
        String keys[] = {"the", "a", "there", "answer", "any",
                         "by", "bye", "their"};

        String output[] = {"Not present in trie", "Present in trie"};


        root = new TrieNode();

        // Construct trie
        int i;
        for (i = 0; i < keys.length ; i++)
            insert(keys[i]);

        // Search for different keys
        if(search("the") == true)
            System.out.println("the --- " + output[1]);
        else System.out.println("the --- " + output[0]);

        if(search("these") == true)
            System.out.println("these --- " + output[1]);
        else System.out.println("these --- " + output[0]);

        if(search("their") == true)
            System.out.println("their --- " + output[1]);
        else System.out.println("their --- " + output[0]);

        if(search("thaw") == true)
            System.out.println("thaw --- " + output[1]);
        else System.out.println("thaw --- " + output[0]);
    }
}
